test/wit.js

```diff
@@ -0,0 +1,158 @@
+import { ok, deepStrictEqual } from "node:assert";
+import { readFile, rm, writeFile, mkdtemp } from "node:fs/promises";
+import { createServer } from "node:http";
+import { tmpdir } from "node:os";
+import { normalize, resolve, sep } from "node:path";
+
+import { fileURLToPath, pathToFileURL } from "url";
+import { HTTPServer } from "@bytecodealliance/preview2-shim/http";
+
+import { componentNew, preview1AdapterCommandPath, componentEmbed, transpile } from "../src/api.js";
+import { exec, jcoPath, getTmpDir } from "./helpers.js";
+
+export async function witTest() {
+  suite("WIT", () => {
+    var tmpDir;
+    var outFile;
+
+    // Content of test/fixtures/wits/feature-gates.wit
+    var featureGatesWitContent;
+
+    suiteSetup(async function () {
+      tmpDir = await getTmpDir();
+      outFile = resolve(tmpDir, "out-component-file");
+      featureGatesWitContent = await readFile(
+        `test/fixtures/wits/feature-gates.wit`,
+        "utf8"
+      );
+    });
+
+    suiteTeardown(async function () {
+      try {
+        await rm(tmpDir, { recursive: true });
+      } catch {}
+    });
+
+    teardown(async function () {
+      try {
+        await rm(outFile);
+      } catch {}
+    });
+
+    // feature-gates: unstable should not be present without all features enabled
+    test("Feature gates", async () => {
+      // Build a dummy WIT component
+      const generatedComponent = await componentEmbed({
+        witSource: featureGatesWitContent,
+        dummy: true,
+        metadata: [
+          ["language", [["javascript", ""]]],
+          ["processed-by", [["dummy-gen", "test"]]],
+        ],
+        features: {
+          tag: "list",
+          val: [
+            "some-feature",
+          ],
+        },
+      });
+      const component = await componentNew(generatedComponent);
+
+      // Transpile the component
+      const { files, imports, exports } = await transpile(component);
+      deepStrictEqual(imports, [
+        "test:feature-gates/foo",
+        ]);
+      deepStrictEqual(exports, [
+        ["foo", "instance"],
+        ["test:feature-gates/foo@0.2.1", "instance"],
+      ]);
+      ok(files['component.js'], "component js was generated");
+      ok(files['component.d.ts'], "component typings were generated");
+      ok(files['interfaces/test-feature-gates-foo.d.ts'], "interface typings were generated");
+
+      // Check the interfaces file for the right exports
+      const interfaces = Buffer.from(files['interfaces/test-feature-gates-foo.d.ts']).toString('utf8');
+      ok(interfaces.includes("export function a(): void;"), "unconstrained export foo/a is present");
+      ok(interfaces.includes("export function b(): void;"), "@since(0.2.1) export foo/b is present (version matches)");
+      ok(interfaces.includes("export function c(): void;"), "@since(0.2.1) export foo/c is present (feature is ignored)");
```

Comment:
This one feels wrong to me - when a feature is not enabled, even if the version matches it should surely still be a requirement that the feature is explicitly enabled?

Reply:
Oh and just to note here -- I'll bring this up in the upcoming CM meeting